Add test_make_fx_model_train example (#980)

mostafaelhoushi opened the pull request:
Summary: Pull Request resolved: https://github.com/pytorch/functorch/pull/980

Test Plan: CI should pass

Differential Revision: D38078694

Pulled By: mostafaelhoushi

mostafaelhoushi commented:
Removing the example that calls `optim.step()` that throws an error. Created a GitHub Issue for it: https://github.com/pytorch/functorch/issues/981

Chillee commented on functorch/test/test_pythonkey.py:
```diff
@@ -74,6 +76,53 @@ def f(x):
+        input = torch.randn(3, 5, requires_grad=True)
+        params = dict(model.named_parameters())
+        fx_f = make_fx(f)(input, params)
+        # TODO: what assert statement should we add here?
```
You should run the traced graph with inputs and compare the output against running the original function.

mostafaelhoushi replied:
Done. `fx_f` may change the order of parameters, and may have a numerical difference, so I took that into consideration when comparing

Chillee reviewed (commented):
Also, can we move this into `pytorch/test/test_proxy_tensor.py`? Thanks!

mostafaelhoushi committed:
Add test_make_fx_model_train example (#82011)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82011

X-link: https://github.com/pytorch/functorch/pull/980

Test Plan: CI should pass

Reviewed By: benoitsteiner

Differential Revision: D38078694

Pulled By: mostafaelhoushi

fbshipit-source-id: 147bef529daf46d68ac4ef0e3fc33cf1f67c5afa

Chillee reviewed (approved):
LGTM now.

mostafaelhoushi committed:
compare different combinations of elements

mostafaelhoushi committed:
resolve conflicts

mostafaelhoushi committed:
resolving conflicts

mostafaelhoushi committed:
add empty line

mostafaelhoushi committed:
Merge branch 'master' into export-D38078694

mostafaelhoushi committed:
use more conservative comparison for model_fwd_bwd

and add comments

mostafaelhoushi commented on test/test_proxy_tensor.py:
```diff
@@ -274,6 +276,72 @@ def fn(x):
+        fx_f = make_fx(f)(input, params, buffers)
+        # fx may change the order of parameters in list, so using set() to compare
+        # also there is a numerical difference in results so changing atol from 1e-08 to 1e-03
+        self.assertTrue(
+            torch.allclose(fx_f(input, params, buffers)[0], f(input, params, buffers)[0], atol=1e-03)
+            or
+            torch.allclose(fx_f(input, params, buffers)[0], f(input, params, buffers)[1], atol=1e-03)
+        )
+        self.assertTrue(
+            torch.allclose(fx_f(input, params, buffers)[1], f(input, params, buffers)[0], atol=1e-03)
+            or
+            torch.allclose(fx_f(input, params, buffers)[1], f(input, params, buffers)[1], atol=1e-03)
+        )
```
Modified `atol` from the default 1e-08 to 1e-03 because there are some numeric differences between `fx_f` and `f`:

```
$ print(fx_f(input, params, buffers)[0])
> tensor([ 0.1805, -0.3061, -0.3142, -0.4370,  0.0827], grad_fn=<SubBackward0>)

$ print(fx_f(input, params, buffers)[1])
> tensor([[-0.0426,  0.4038, -0.3666, -0.4284, -0.3646],
        [-0.0428,  0.0160,  0.3607, -0.1020, -0.0766],
        [-0.2216, -0.0094,  0.2439,  0.0439, -0.4262],
        [-0.4087,  0.2085,  0.2881,  0.2005,  0.2684],
        [ 0.0596, -0.4240, -0.2393,  0.0553, -0.1873]], grad_fn=<SubBackward0>)

$ print(f(input, params, buffers)[0])
> tensor([[-0.0421,  0.4038, -0.3665, -0.4285, -0.3645],
        [-0.0429,  0.0167,  0.3603, -0.1018, -0.0764],
        [-0.2218, -0.0087,  0.2435,  0.0441, -0.4261],
        [-0.4084,  0.2090,  0.2881,  0.1996,  0.2684],
        [ 0.0604, -0.4227, -0.2395,  0.0545, -0.1872]], grad_fn=<SubBackward0>)

$ print(f(input, params, buffers)[1])
> tensor([ 0.1800, -0.3066, -0.3147, -0.4375,  0.0812], grad_fn=<SubBackward0>)
```

mostafaelhoushi commented:
@pytorchbot merge
